src/ILLink.Shared/Dataflow/DictionaryLattice.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+namespace ILLink.Shared
+{
+	public class DefaultValueDictionary<TKey, TValue> : Dictionary<TKey, TValue>,
+ 		IEquatable<DefaultValueDictionary<TKey, TValue>>
+		where TKey : IEquatable<TKey>
+		where TValue : IEquatable<TValue>
+	{
+		TValue DefaultValue { get; }
+
+		public DefaultValueDictionary (TValue defaultValue) => DefaultValue = defaultValue;
+
+		public DefaultValueDictionary (DefaultValueDictionary<TKey, TValue> other)
+			: base (other)
+		{
+			DefaultValue = other.DefaultValue;
+		}
+
+		public TValue Get (TKey key) => TryGetValue (key, out var value) ? value : DefaultValue;
+
+		public void Set (TKey key, TValue value)
```

Comment:
This is tricky - if I forget and call `Add` or use the indexer, it will actually insert the default value. Functionality-vise it's still going to work just fine, but there will effectively be a "memory leak" (well sort of).

I wonder if it would be better to simply wrap the dictionary and have only a limitted APIs available on the class. It's unlikely we will need it to act as a true dictionary anyway.

Reply:
Good point, I made it a wrapper struct instead.